src/smt/optimization_solver.h

```diff
@@ -80,38 +90,140 @@ class Objective
+  /** Maximize the target node with constraints encoded in parentSMTSolver 
+   * Parameters: 
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints 
+   * - target: the target expression to optimize 
+   * Return value: 
+   * - std::pair<OptResult, CVC4::Node>: the result of optimization and the optimized value 
+   **/
+  virtual std::pair<OptResult, CVC4::Node> maximize(SmtEngine *parentSMTSolver, CVC4::Node target) = 0;
+};
+
+/**
+ * Optimizer for Integer datatype 
+ */
+struct OMTOptimizerInteger : OMTOptimizer 
+{
+public:
+  virtual ~OMTOptimizerInteger() = default;
+  /** Minimize the target node with constraints encoded in parentSMTSolver 
+   * Parameters: 
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints 
+   * - target: the target expression to optimize 
+   * Return value: 
+   * - std::pair<OptResult, CVC4::Node>: the result of optimization and the optimized value 
+   **/
+  virtual std::pair<OptResult, CVC4::Node> minimize(SmtEngine *parentSMTSolver, CVC4::Node target) override;
+  /** Maximize the target node with constraints encoded in parentSMTSolver 
+   * Parameters: 
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints 
+   * - target: the target expression to optimize 
+   * Return value: 
+   * - std::pair<OptResult, CVC4::Node>: the result of optimization and the optimized value 
+   **/
+  virtual std::pair<OptResult, CVC4::Node> maximize(SmtEngine *parentSMTSolver, CVC4::Node target) override;
+private:
+  /** Handles the optimization query specified by objType (=OBJECTIVE_MINIMIZE/MAXIMIZE) **/
+  std::pair<OptResult, CVC4::Node> optimize(SmtEngine *parentSMTSolver, CVC4::Node target, ObjectiveType objType);
+};
+
+/**
+ * Optimizer for BitVector datatype 
+ */
+struct OMTOptimizerBitVector : OMTOptimizer 
+{
+public:
+  OMTOptimizerBitVector(bool isSigned);
+  virtual ~OMTOptimizerBitVector() = default;
+  /** Minimize the target node with constraints encoded in parentSMTSolver 
+   * Parameters: 
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints 
+   * - target: the target expression to optimize 
+   * Return value: 
+   * - std::pair<OptResult, CVC4::Node>: the result of optimization and the optimized value 
+   **/
+  virtual std::pair<OptResult, CVC4::Node> minimize(SmtEngine *parentSMTSolver, CVC4::Node target) override;
+  /** Maximize the target node with constraints encoded in parentSMTSolver 
+   * Parameters: 
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints 
+   * - target: the target expression to optimize 
+   * Return value: 
+   * - std::pair<OptResult, CVC4::Node>: the result of optimization and the optimized value 
+   **/
+  virtual std::pair<OptResult, CVC4::Node> maximize(SmtEngine *parentSMTSolver, CVC4::Node target) override;
+private:
+  /** Computes the BitVector version of (a + b) / 2 without overflow, rounding towards -infinity **/
```

Comment:
can u clarify in the documentation what do you mean by rounding towards -infinity? And also can u document where is this being used?

Reply:
rounding towards -inf means -1.5 ≈ -2 
rounding towards 0 means -1.5 ≈ -1

Reply:
great, please clarify inside the comment in the code.